### PR TITLE
Add ability to use $null: false and $notNull: false

### PR DESCRIPTION
**Pull request by petersg83**
In the REST API, we could filter with 
```
filters: {
  name: {
    $null: true,
  }
}

// and

filters: {
  name: {
    $notNull: true,
  }
}
```

but not 
```
filters: {
  name: {
    $null: false,
  }
}

// and

filters: {
  name: {
    $notNull: false,
  }
}
```

`false` value was not taken into account. The PR aims to allow the false value for `$null` and `$notNull`

### CODE REVIEW

**Review by Convly — DISMISSED** [review]
LGTM

File: packages/core/database/lib/utils/boolean.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
+'use strict';
+
+const parseBoolean = value => {
```
Comment (alexandrebodin) [inline packages/core/database/lib/utils/boolean.js:3]: we export this function from @strapi/utils I believe. if you are moving it out I guess better just import it from the strapi utils ?

File: packages/core/database/lib/query/helpers/where.js [hunk 2]
```diff
@@ -247,14 +248,18 @@ const applyOperator = (qb, column, operator, value) => {
       break;
     }
     case '$null': {
-      if (value) {
+      if (parseBoolean(value)) {
```
Comment (alexandrebodin) [inline packages/core/database/lib/query/helpers/where.js:251]: I really don't think this makes sense to do this parsing here. I think this should be part of the convertFilters function in the convert query params phase. I don't see a good reason to allow this here

**Review by alexandrebodin — CHANGES_REQUESTED** [review]
Make sure to add a description in the PR following the template (and add labels + milestone) :)

File: packages/core/database/lib/utils/boolean.js (hunk 1 above)
Comment (petersg83) [reply packages/core/database/lib/utils/boolean.js:3]: Yes we can, however it is not exactly the same.
In `@strapi/utils` it is:
```js
const parseType = ({ type, value }) => {
  switch (type) {
    case 'boolean': {
      if (typeof value === 'boolean') return value;

      if (['true', 't', '1', 1].includes(value)) {
        return true;
      }

      if (['false', 'f', '0', 0].includes(value)) {
        return false;
      }

      throw new Error('Invalid boolean input. Expected "t","1","true","false","0","f"'); // <- here is the difference
    }
    case 'integer':
    case 'biginteger':
    case 'float':
    case 'decimal': {
      return _.toNumber(value);
    }
    case 'time': {
      return parseTime(value);
    }
    case 'date': {
      return parseDate(value);
    }
    case 'timestamp':
    case 'datetime': {
      return parseDateTimeOrTimestamp(value);
    }
    default:
      return value;
  }
};
```

I can use it but we need to decide if we throw an error or if we cast to boolean anyway. What do you think?

File: packages/core/database/lib/query/helpers/where.js (hunk 2 above)
Comment (petersg83) [reply packages/core/database/lib/query/helpers/where.js:251]: I'm not sure where you are referring to. Would you have 5 minutes to discuss about it?

Comment (alexandrebodin) [reply packages/core/database/lib/query/helpers/where.js:251]: convertFiltersQueryParams in packages/core/utils/lib/convert-query-params.js. right now it's not doing anything. that's where we would do these kind of casting. to avoid having that at the db level

Comment (alexandrebodin) [reply packages/core/database/lib/query/helpers/where.js:251]: The code is already handling false but not 'false'.
I think it's better that way as it's at a programmatic level. 
But we should still be able to handle 'false', 'f', 't', 'true' at a higher level instead (ideally it would be at controller/resolver) level, but in a first iteration we can handle it at the entityService level

File: packages/core/database/lib/utils/boolean.js (hunk 1 above)
Comment (alexandrebodin) [reply packages/core/database/lib/utils/boolean.js:3]: We don't want to make a breaking change so it depends what we currently do. I think we won't be able to throw for now. However we can split part of the logic to have the parseBoolean function and only throw in parseType

File: packages/core/database/lib/query/helpers/where.js (hunk 2 above)
Comment (petersg83) [reply packages/core/database/lib/query/helpers/where.js:251]: Yes I thought that way to and then I saw that in `packages/core/database/lib/fields.js`, we parse booleans for a field (toDB). I was wondering if we should apply the same logic for the filters parsing and the fields parsing. Does it make sense to parse for the fields bu not the filters in the package Database? I think it can make sense but I would like a confirmation

Comment (alexandrebodin) [reply packages/core/database/lib/query/helpers/where.js:251]: I think I would move that out of the DB too now.

File: packages/core/utils/lib/convert-query-params.js [hunk 3]
```diff
@@ -250,39 +251,52 @@ const sanitizeFilters = (filters, schema) => {
...
-    else if (isObject(value)) {
-      filters[key] = sanitizeFilters(value, schema);
+    else {
+      if (isObject(value)) {
+        filters[key] = convertAndSanitizeFilters(value, schema);
+      } else if (['$null', '$notNull'].includes(key)) {
+        try {
+          filters[key] = parseType({ type: 'boolean', value: filters[key], forceCast: true });
+        } catch (e) {
+          throw new ApplicationError(e.message);
```
Comment (alexandrebodin) [inline packages/core/utils/lib/convert-query-params.js:297]: WHat do you think of using a more details error format like we do for Validation Error ? We could have QueryError or similar ?

**Review by alexandrebodin — COMMENTED** [review]
Starting to look good :D

File: packages/core/utils/lib/convert-query-params.js (hunk 3 above)
Comment (petersg83) [reply packages/core/utils/lib/convert-query-params.js:297]: I saw that the error handling can be improved, for example there is already a 500 error when we select a field that doesn't exist.
I highly hesitated to catch the error or to just cast any value there to Boolean. In anycase it doesn't fit a proper error handling system. It would be nice to have one, I thought that we could do it in another US and that we should maybe discuss about it before. What do you think ?

File: packages/core/utils/lib/convert-query-params.js [hunk 4]
```diff
@@ -250,39 +251,52 @@ const sanitizeFilters = (filters, schema) => {
...
 
     // Handle operators
-    else if (isObject(value)) {
-      filters[key] = sanitizeFilters(value, schema);
+    else {
+      if (isObject(value)) {
+        filters[key] = convertAndSanitizeFilters(value, schema);
+      } else if (['$null', '$notNull'].includes(key)) {
+        try {
+          filters[key] = parseType({ type: 'boolean', value: filters[key], forceCast: true });
```
Comment (alexandrebodin) [inline packages/core/utils/lib/convert-query-params.js:295]: I don't see forceCast as an option in the code

File: packages/core/utils/lib/convert-query-params.js (hunk 3 above)
Comment (alexandrebodin) [reply packages/core/utils/lib/convert-query-params.js:297]: Let's discuss it during our next sync :)

File: packages/core/strapi/tests/filtering.test.e2e.js [hunk 5]
```diff
@@ -202,13 +202,23 @@ describe('Filtering API', () => {
     });
 
     describe('Filter null', () => {
-      test('Should return only matching items', async () => {
+      test.each([
```
Comment (alexandrebodin) [inline packages/core/strapi/tests/filtering.test.e2e.js:205]: Can you add tests for edge cases like : 
```js
''
[]
{}
```

**Review by alexandrebodin — CHANGES_REQUESTED** [review]

File: packages/core/strapi/tests/filtering.test.e2e.js (hunk 5 above)
Comment (petersg83) [reply packages/core/strapi/tests/filtering.test.e2e.js:205]: yes :)

File: packages/core/strapi/tests/filtering.test.e2e.js [hunk 6]
```diff
@@ -220,15 +233,26 @@ describe('Filtering API', () => {
...
       });
 
-      test('Should return three matches', async () => {
+      test.each([
+        [{ $notNull: true }],
+        [{ $notNull: 'true' }],
+        [{ $notNull: '1' }],
+        [{ $notNull: 't' }],
+        [{ $notNull: 'anything' }],
+        [{ $notNull: '[]' }],
```
Comment (alexandrebodin) [inline packages/core/strapi/tests/filtering.test.e2e.js:242]: Any reason not to test against a realy array not a string ? this is something we can receive (same remark for objects)

Comment (petersg83) [reply packages/core/strapi/tests/filtering.test.e2e.js:242]: Mmmh it's a bit tricky here.
- `[]` and `{}` is not stringified by `qs` as it is empty, so it is not present in the query string and cannot be tested here
- However the query can be manually set (by the user) to `filters[code][$null]=`, `qs` will then parse it as `{ filters: { code: { '$null': '' } } }`, not to an empty array of empty object. We already handle the `''` usecase.
- The query can be a non-empty object or array, in this case I can parse it as truthy, will do

**Review by alexandrebodin — APPROVED** [review]
